mangaki/mangaki/utils/recommendations.py

```diff
@@ -5,119 +5,142 @@
+
+    interesting_works = set()
+    for work_id, _ in nb_ratings.most_common():
+        if nb_ratings[work_id] < RATED_BY_AT_LEAST:
+            break
+        interesting_works.add(work_id)
+
+    for user_id, work_id, rating in triplets:
+        X.append((encode_user[user_id], encode_work[work_id]))
+        y.append(rating_values[rating])
+    return np.array(X), np.array(y), len(users), len(works), encode_user, decode_user, encode_work, decode_work, interesting_works
+
+
+def get_reco_algo(user, algo_name='knn', category='all'):
+    chrono = Chrono(CHRONO_ENABLED)
 
+    already_rated_works = Rating.objects.filter(user=user).values_list('work_id', flat=True)
 
-NB_NEIGHBORS = 15
-MIN_RATINGS = 3
+    chrono.save('[%dQ] get rated works' % len(connection.queries))
 
-CHRONO_ENABLED = False
+    if algo_name == 'knn':
+        queryset = Rating.objects.filter(work__in=already_rated_works)
+        ratings_pack = make_anonymous_data(queryset)
 
+        chrono.save('[%dQ] make first anonymous data' % len(connection.queries))
 
-def get_recommendations(user, category, editor='unspecified', dpp=False):
-    # What if user is not authenticated? We will see soon.
-    chrono = Chrono(CHRONO_ENABLED)
+        X, y, nb_users, nb_works, encode_user, decode_user, encode_work, decode_work, interesting_works = ratings_pack
+        algo = MangakiKNN(NB_NEIGHBORS, single_user=True, missing_is_mean=False)
+        algo.set_parameters(nb_users, nb_works)
+        algo.fit(X, y)
 
-    chrono.save('[%dQ] begin' % len(connection.queries))
+        chrono.save('[%dQ] prepare first fit' % len(connection.queries))
 
-    rated_works = {}
-    if dpp:
-        qs = ColdStartRating.objects.filter(user=user).values_list('work_id', 'choice')
-        values = rating_values_dpp
+        algo.get_neighbors([encode_user[user.id]])
+        encoded_neighbors = list(algo.closest_neighbors[encode_user[user.id]].items())
+        neighbors = list(map(lambda x: decode_user[x[0]], encoded_neighbors))
 
+        chrono.save('[%dQ] get neighbors (checksum: %d)' % (len(connection.queries), sum(neighbors)))
+
+        queryset = Rating.objects.filter(user__id__in=neighbors + [user.id]).exclude(choice__in=['willsee', 'wontsee'])
     else:
-        qs = Rating.objects.filter(user=user).values_list('work_id', 'choice')
-        values = rating_values
-
-    for work_id, choice in qs:
-        rated_works[work_id] = choice
-
-    willsee = set()
-    if user.profile.reco_willsee_ok:
-        banned_works = set()
-        for work_id in rated_works:
-            if rated_works[work_id] != 'willsee':
-                banned_works.add(work_id)
-            else:
-                willsee.add(work_id)
+        queryset = Rating.objects.all()
+
+    chrono.save('[%dQ] get interesting %d ratings' % (len(connection.queries), queryset.count()))
+
+    filename = 'ratings.pickle'
+    if algo_name == 'knn' or not os.path.isfile('ratings.pickle'):
```

Comment:
Could we detail why do we make an exception for `knn` when saving the ratings?
Also, could we prefer another format rather than `pickle` for the export (e.g. JSON).